### PR TITLE
[9] Revamp Thread-Tab of Forum

### PR DESCRIPTION
**Pull request by gvollbach**
https://docu.ilias.de/goto_docu_wiki_wpage_7529_1357.html

### CODE REVIEW

File: Modules/Forum/classes/class.ilForumThreadObject.php [hunk 1]
```diff
@@ -0,0 +1,146 @@
+<?php
+
+class ilForumThreadObject extends ilTable2GUI
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilForumThreadObject.php]: Please rename this by using a `TableGUI` suffix.

File: Modules/Forum/classes/class.ilForumExportGUI.php [hunk 2]
```diff
@@ -374,14 +374,12 @@ public function exportHTML(): void
                 'thread_ids',
                 $this->refinery->kindlyTo()->listOf($this->refinery->kindlyTo()->int())
             );
-        } else {
-            if($this->ref_id > 0) {
+        } else if($this->ref_id > 0) {
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilForumExportGUI.php]: `elseif` :-)

**Comment by mjansenDatabay** [comment]
@gvollbach Please squash the commits to 1 single commit. Thx.

File: Modules/Forum/classes/class.ilObjForumGUI.php [hunk 3]
```diff
@@ -765,72 +769,432 @@ public function getContent(): string
...
+            $this->lng->txt("frm_edit_title"),
+            null,
+            [
+                $this->factory->input()->field()->text($this->lng->txt("frm_edit_title"))->withValue($topic->getSubject()),
+            ],
+            $target
+        );
+    }
+
+    private function addDraftButtonIfDraftsExists($cmd)
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilObjForumGUI.php]: Missing types

File: Modules/Forum/classes/class.ilForumThreadObjectTableGUI.php [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+ *
+ * If this is not the case or you just want to try ILIAS, you'll find
+ * us at:
+ * https://www.ilias.de
+ * https://github.com/ILIAS-eLearning
+ *
+ *********************************************************************/
+
+declare(strict_types=1);
+class ilForumThreadObjectTableGUI extends ilTable2GUI
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilForumThreadObjectTableGUI.php:21]: Please add a blank line

File: Modules/Forum/classes/class.ilForumThreadObjectTableGUI.php [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        $this->setDefaultOrderDirection('DESC');
+        $this->setDefaultOrderField('lp_date');
+        $this->setExternalSorting(true);
+        $this->setExternalSegmentation(true);
+
+        parent::__construct($a_parent_obj, $a_parent_cmd);
+        $this->mainTemplate->addCss('./Modules/Forum/css/forum_table.css');
+        $this->is_post_draft_allowed = ilForumPostDraft::isSavePostDraftAllowed();
+    }
+    public function fetchDataAnReturnObject(): array
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilForumThreadObjectTableGUI.php:65]: Please add a blank line

File: Modules/Forum/classes/class.ilObjForumGUI.php [hunk 6]
```diff
@@ -765,72 +769,432 @@ public function getContent(): string
...
+        $this->renderThreadOverview($tbl, $frm, $frm_object);
+        return '';
+    }
+
+    /**
+     * @param ilForumThreadObjectTableGUI $tbl
+     * @param ilForum                     $frm
+     * @param ForumDto                    $frm_object
+     * @return void
+     * @throws ilCtrlException
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilObjForumGUI.php]: Please remove obsolete information

File: Modules/Forum/classes/class.ilObjForumGUI.php [hunk 7]
```diff
@@ -765,72 +769,432 @@ public function getContent(): string
...
+        $offset = 0;
+        if ($this->http->wrapper()->query()->has('page')) {
+            $offset = $this->http->wrapper()->query()->retrieve('page', $this->refinery->kindlyTo()->int());
+        }
+        return $offset;
+    }
+
+    /**
+     * @param ilForumThreadObjectTableGUI $tbl
+     * @return void
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilObjForumGUI.php]: Please remove obsolete information

File: Modules/Forum/classes/class.ilObjForumGUI.php [hunk 8]
```diff
@@ -765,72 +769,432 @@ public function getContent(): string
...
+        $limit = ilForumProperties::PAGE_SIZE_THREAD_OVERVIEW;
+        $offset = $this->getThreadOffset() * $limit;
+
+        $tbl->setOffset($offset);
+        $tbl->setLimit($limit);
+    }
+
+    /**
+     * @param ilForumThreadObjectTableGUI $tbl
+     * @return void
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilObjForumGUI.php]: Please remove obsolete information

File: Modules/Forum/classes/class.ilObjForumGUI.php [hunk 9]
```diff
@@ -765,72 +769,432 @@ public function getContent(): string
...
+            $tbl->setOrderField($sortation_options[$sortation_value]['field']);
+        } else {
+            $tbl->setOrderDirection($sortation_options[1]['direction']);
+            $tbl->setOrderField($sortation_options[1]['field']);
+        }
+    }
+
+    /**
+     * @return array
+     * @throws ilCtrlException
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilObjForumGUI.php]: Please try to specify the shape of the array

File: Modules/Forum/classes/class.ilObjForumGUI.php [hunk 10]
```diff
@@ -765,72 +769,432 @@ public function getContent(): string
...
+                'thr_pk',
+                $this->refinery->kindlyTo()->listOf($this->refinery->kindlyTo()->int())
+            );
+        }
+
+        return $thread_ids;
+    }
+
+    /**
+     * @param ilForumTopic $forum_topic
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilObjForumGUI.php]: Please remove obsolete information

File: Modules/Forum/classes/class.ilObjForumGUI.php [hunk 11]
```diff
@@ -765,72 +769,432 @@ public function getContent(): string
...
+        $this->ctrl->setParameter($this, 'page', null);
+
+        return $f->button()->shy($title, $url);
+    }
 
-            $ref_id = $this->retrieveRefId();
+    /**
+     * @param int          $ref_id
+     * @param ilForumTopic $forum_topic
+     * @return Standard
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilObjForumGUI.php]: Please remove obsolete information

File: Modules/Forum/classes/class.ilObjForumGUI.php [hunk 12]
```diff
@@ -765,72 +769,432 @@ public function getContent(): string
...
+                $move,
+                $merge,
+                $delete
+            ]);
+        }
+
+        return $actions;
+    }
+
+    private function getEditTitleModal(ilForumTopic $topic)
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilObjForumGUI.php]: The return type is missing here

File: Modules/Forum/classes/class.ilObjForumGUI.php [hunk 13]
```diff
@@ -765,72 +769,432 @@ public function getContent(): string
...
+
+            return $draft_modal;
         }
+    }
 
-        $this->tpl->setPermanentLink($this->object->getType(), $this->object->getRefId(), '', '_top');
+    /**
+     * @param ilForumTopic $forum_topic
+     * @param int          $ref_id
+     * @return Shy
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilObjForumGUI.php]: Please remove obsolete information

File: Modules/Forum/classes/class.ilObjForumGUI.php [hunk 14]
```diff
@@ -765,72 +769,432 @@ public function getContent(): string
...
+        }
+        return $stick_or_no_stick;
+    }
 
-        $this->initStyleSheets();
+    /**
+     * @param ilForumTopic $forum_topic
+     * @param int          $ref_id
+     * @return Shy
+     * @throws ilCtrlException
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilObjForumGUI.php]: Please remove obsolete information

File: Modules/Forum/classes/class.ilObjForumGUI.php [hunk 15]
```diff
@@ -4457,6 +4826,7 @@ public function mergeThreadsObject(): void
         }
 
         $threadIdToMerge = (int) ($this->httpRequest->getQueryParams()['merge_thread_id'] ?? 0);
+        $threadIdToMerge = (int) ($this->http->wrapper()->query()->retrieve('thr_pk', $this->refinery->kindlyTo()->int()) ?? 0);
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilObjForumGUI.php]: This makes not sense. The transformation either returns an integer, or raises an exception. NULL is not possible.

Please try this.

```php
$threadIdToMerge = $this->http->wrapper()->query()->retrieve(
    'thr_pk',
    $this->refinery->byTrying([
        $this->refinery->kindlyTo()->int(),
        $this->refinery->always(0)
    ])
);
```

File: Modules/Forum/classes/class.ilObjForumGUI.php [hunk 16]
```diff
@@ -765,72 +769,432 @@ public function getContent(): string
...
+            $list_item = $list_item->withColor($df->color('#f3de2c'));
+        }
+        return $list_item;
+    }
 
-        $this->tpl->setContent($forwarder->forward() . $threadsTemplate->get());
+    /**
+     * @param ilForumTopic                  $current_thread
+     * @param Item $list_item
+     * @return Item
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilObjForumGUI.php]: Please remove obsolete information

File: Modules/Forum/classes/class.ilObjForumGUI.php [hunk 17]
```diff
@@ -765,72 +769,432 @@ public function getContent(): string
...
-            $this->object,
-            $this->user,
-            $this->content_style_domain
-        );
-        $forwarder->setPresentationMode(ilForumPageCommandForwarder::PRESENTATION_MODE_PRESENTATION);
+    /**
+     * @param Item $list_item
+     * @param int  $author_id
+     * @param int  $ref_id
+     * @return Item
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilObjForumGUI.php]: Please remove obsolete information

**Review by mjansenDatabay — CHANGES_REQUESTED** [review]

**Comment by gvollbach** [comment]
> @gvollbach Please squash the commits to 1 single commit. Thx.

Done.

File: Modules/Forum/classes/class.ilForumExportGUI.php [hunk 18]
```diff
@@ -374,6 +374,12 @@ public function exportHTML(): void
                 'thread_ids',
                 $this->refinery->kindlyTo()->listOf($this->refinery->kindlyTo()->int())
             );
+        } elseif($this->ref_id > 0) {
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilForumExportGUI.php]: Whitespace, again ^^. I again suggest to use the `Captain` to run the PHP-CS-FIXER on every commit :-).

File: lang/ilias_de.lang [hunk 19]
```diff
@@ -9731,7 +9739,7 @@ forum#:#mark_moderator_posts_desc#:#Beiträge der Moderation werden farblich her
 forum#:#merge#:#Zusammenführen
 forum#:#merge_posts_into_thread#:#Verschmelze Posts mit anderem Thema
 forum#:#merged_threads_successfully#:#Die selektierten Themen wurden erfolgreich zusammengeführt
```
Comment (oliversamoila) [inline lang/ilias_de.lang:9741]: Die Themen wurden erfolgreich zusammengeführt.

File: Modules/Forum/classes/class.ilForumThreadObjectTableGUI.php (hunk 4 above)
Comment (mjansenDatabay) [reply Modules/Forum/classes/class.ilForumThreadObjectTableGUI.php:21]: No, this is not solved.

File: Modules/Forum/classes/class.ilForumThreadObjectTableGUI.php (hunk 5 above)
Comment (mjansenDatabay) [reply Modules/Forum/classes/class.ilForumThreadObjectTableGUI.php:65]: No, this is not solved.

File: Modules/Forum/classes/class.ilObjForumGUI.php (hunk 11 above)
Comment (mjansenDatabay) [reply Modules/Forum/classes/class.ilObjForumGUI.php]: No, this is not solved.

**Review by mjansenDatabay — CHANGES_REQUESTED** [review]
Please check my new inline comments.

**Review by mjansenDatabay — APPROVED** [review]

**Pull request merged**